### ISSUE TITLE
Compile error when cross compiling

### ISSUE THREAD

**Issue by MylesAndMore** [karlseguin/http.zig#55]
I should prefix this issue by noting that the error here seems to stem from the `metrics.zig` library (or possibly the usage of said library, I'm not sure). If this isn't the right place to post this issue, I'm happy to move it!

When trying to cross compile my project using http.zig I encountered a compiler error (see below). This error is also present in the example and is *not* present when building a native binary.

The target is `arm-linux-gnueabihf` (a Raspberry Pi) with CPU `arm1176jzf_s` (although this happens with/without the CPU flag), my host is `x86_64-linux-gnueabihf`, and I'm using Zig 0.13 (as well as the respective http.zig branch). Let me know if I need to provide any further details.

```
http.zig [zig-0.13] zig build run -Dtarget=arm-linux-gnueabihf -Dcpu=arm1176jzf_s     
run
└─ run http.zig demo
   └─ zig build-exe http.zig demo Debug arm-linux-gnueabihf 2 errors
/home/myles/.cache/zig/p/1220232ab38d0c2cfb10680115c17ad2fa1a8531dbaf8bbfb359ec67e80c7d5f5758/src/counter.zig:70:31: error: expected 32-bit integer type or smaller; found 64-bit integer type
    const count = @atomicLoad(V, &self.count, .monotonic);
                              ^
referenced by:
    write__anon_11135: src/metrics.zig:63:41
    writeMetrics__anon_11134: src/httpz.zig:37:40
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
/home/myles/.cache/zig/p/1220232ab38d0c2cfb10680115c17ad2fa1a8531dbaf8bbfb359ec67e80c7d5f5758/src/counter.zig:65:20: error: expected 32-bit integer type or smaller; found 64-bit integer type
    _ = @atomicRmw(V, &self.count, .Add, count, .monotonic);
                   ^
error: the following command failed with 2 compilation errors:
/usr/bin/zig build-exe -ODebug -target arm-linux-gnueabihf -mcpu arm1176jzf_s --dep httpz --dep metrics --dep websocket -Mroot=/home/myles/http.zig/example/main.zig --dep metrics --dep websocket --dep build -Mhttpz=/home/myles/http.zig/src/httpz.zig -Mmetrics=/home/myles/.cache/zig/p/1220232ab38d0c2cfb10680115c17ad2fa1a8531dbaf8bbfb359ec67e80c7d5f5758/src/metrics.zig -Mwebsocket=/home/myles/.cache/zig/p/12208720b772330f309cfb48957f4152ee0930b716837d0c1d07fee2dea2f4dc712e/src/websocket.zig -Mbuild=/home/myles/http.zig/.zig-cache/c/0a991cbb44bf1c4520ed99cc54c5dead/options.zig --cache-dir /home/myles/http.zig/.zig-cache --global-cache-dir /home/myles/.cache/zig --name http.zig demo --listen=- 
Build Summary: 1/6 steps succeeded; 1 failed (disable with --summary none)
run transitive failure
└─ run http.zig demo transitive failure
   ├─ zig build-exe http.zig demo Debug arm-linux-gnueabihf 2 errors
   └─ install transitive failure
      └─ install http.zig demo transitive failure
         └─ zig build-exe http.zig demo Debug arm-linux-gnueabihf (+1 more reused dependencies)
error: the following build command failed with exit code 1:
/home/myles/http.zig/.zig-cache/o/81269789ddbcc896fc2e9fa187e1b17b/build /usr/bin/zig /home/myles/http.zig /home/myles/http.zig/.zig-cache /home/myles/.cache/zig --seed 0x584b3d12 -Ze3649f786d4f72ee run -Dtarget=arm-linux-gnueabihf -Dcpu=arm1176jzf_s
```

Thanks for your help!

**Comment by karlseguin** [comment]
This should now be fixed.

I did the same thing in pg.zig (https://github.com/karlseguin/pg.zig/issues/17), fixed it there but never here.

**Comment by MylesAndMore** [comment]
Great, that seems to have fixed that issue.

However, I've encountered another similar type mismatch; this one comes up when using websockets.
Trying to build the example provided at the bottom of the readme gives:

```
example [zig-0.13] zig build run -Dtarget=arm-linux-gnueabihf 
run
└─ run http.zig demo
   └─ install
      └─ install http.zig demo
         └─ zig build-exe http.zig demo Debug arm-linux-gnueabihf 1 errors
/home/myles/.cache/zig/p/12208720b772330f309cfb48957f4152ee0930b716837d0c1d07fee2dea2f4dc712e/src/reader.zig:156:19: error: expected type 'usize', found 'u64'
      data_needed += payload_length;
      ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
/home/myles/.cache/zig/p/12208720b772330f309cfb48957f4152ee0930b716837d0c1d07fee2dea2f4dc712e/src/reader.zig:156:19: note: unsigned 32-bit int cannot represent all possible unsigned 64-bit values
referenced by:
    readLoop__anon_11867: /home/myles/.cache/zig/p/12208720b772330f309cfb48957f4152ee0930b716837d0c1d07fee2dea2f4dc712e/src/server.zig:282:38
    handle__anon_10932: /home/myles/.cache/zig/p/12208720b772330f309cfb48957f4152ee0930b716837d0c1d07fee2dea2f4dc712e/src/server.zig:183:16
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
error: the following command failed with 1 compilation errors:
/usr/bin/zig build-exe -ODebug -target arm-linux-gnueabihf -mcpu baseline --dep httpz --dep metrics --dep websocket -Mroot=/home/myles/Documents/http.zig/example/main.zig --dep metrics --dep websocket --dep build -Mhttpz=/home/myles/Documents/http.zig/src/httpz.zig -Mmetrics=/home/myles/.cache/zig/p/1220232ab38d0c2cfb10680115c17ad2fa1a8531dbaf8bbfb359ec67e80c7d5f5758/src/metrics.zig -Mwebsocket=/home/myles/.cache/zig/p/12208720b772330f309cfb48957f4152ee0930b716837d0c1d07fee2dea2f4dc712e/src/websocket.zig -Mbuild=/home/myles/Documents/http.zig/.zig-cache/c/0a991cbb44bf1c4520ed99cc54c5dead/options.zig --cache-dir /home/myles/Documents/http.zig/.zig-cache --global-cache-dir /home/myles/.cache/zig --name http.zig demo --listen=- 
Build Summary: 1/6 steps succeeded; 1 failed (disable with --summary none)
run transitive failure
└─ run http.zig demo transitive failure
   ├─ zig build-exe http.zig demo Debug arm-linux-gnueabihf 1 errors
   └─ install transitive failure
      └─ install http.zig demo transitive failure
         └─ zig build-exe http.zig demo Debug arm-linux-gnueabihf (+1 more reused dependencies)
error: the following build command failed with exit code 1:
/home/myles/Documents/http.zig/.zig-cache/o/46fa931df3499313b09784cc80866859/build /usr/bin/zig /home/myles/Documents/http.zig /home/myles/Documents/http.zig/.zig-cache /home/myles/.cache/zig --seed 0xfbd6e5be -Z67283ac9c6fb892c run -Dtarget=arm-linux-gnueabihf
```

**Comment by MylesAndMore** [comment]
I was poking around and got things to work (I think) by changing `reader.zig:156` to
```zig
data_needed += @intCast(payload_length);
```
and `server.zig:232` to
```zig
const l: u64 = data.len;
```
in the `websocket.zig` library. It compiles and basic functionality seems to work, but I haven't tested too much.

There was also a memory leak after running this (when stopping the server without externally closing the socket), although it could be due to my questionable fix, the example, my incompetence, I'm not really sure. Putting it here just in case.
```
error(gpa): memory address 0xaefae000 leaked: 
/home/myles/.cache/zig/p/12208720b772330f309cfb48957f4152ee0930b716837d0c1d07fee2dea2f4dc712e/src/buffer.zig:54:36: 0x117237 in static (nwdrone)
/home/myles/.cache/zig/p/12208720b772330f309cfb48957f4152ee0930b716837d0c1d07fee2dea2f4dc712e/src/reader.zig:50:31: 0x117387 in init (nwdrone)
/home/myles/.cache/zig/p/12208720b772330f309cfb48957f4152ee0930b716837d0c1d07fee2dea2f4dc712e/src/server.zig:178:27: 0xf3127 in handle__anon_13814 (nwdrone)
/home/myles/.cache/zig/p/1220d945f79c6912429cc6757d2944e2ba2311a98bdcef49a1240ba97c41b92da568/src/httpz.zig:665:18: 0xceb87 in websocketHandler__anon_13564 (nwdrone)
/usr/lib/zig/std/Thread.zig:408:13: 0xb7547 in callFn__anon_13069 (nwdrone)
/usr/lib/zig/std/Thread.zig:674:30: 0x9cbb7 in entryFn (nwdrone)
```

**Comment by karlseguin** [comment]
I think `data_needed` needs to become `u64` rather than forcing `payload_length` to `usize`, since (for whatever reason) websocket really does allow 2^64 payloads. Weird though that I can't reproduce this build failure, but I could the other. Anyways, hopefully that's also fixed now.

**Comment by karlseguin** [comment]
As for the memory leak...

websocket.zig is currently being overhauled. It'll be nonblocking. There have been other changes, and one of those changes is cleaner shutdown.

As part of the changes to websocket.zig, httpz is also seeing a smaller set of breaking changes. Part of this is to better integrate with websocket.zig, part of it is just to try to make httpz easier to use.

If you're feeling experimental, you could try to [nonblocking_ws branch](https://github.com/karlseguin/http.zig/tree/nonblocking_ws) of httpz (which uses the [nonblocking branch](https://github.com/karlseguin/websocket.zig/tree/nonblocking) of websocket.zig). There are [migrations guide](https://github.com/karlseguin/http.zig/wiki/Migration) for [both](https://github.com/karlseguin/websocket.zig/wiki/Server-Migration)

**Comment by MylesAndMore** [comment]
> I think `data_needed` needs to become `u64` rather than forcing `payload_length` to `usize`, since (for whatever reason) websocket really does allow 2^64 payloads.

Ah, I didn't know that. It does seem like the better solution, but many of the other functions in `reader` take `usize` arguments which are now incompatible. See another error below (and I think there may possibly be more where that came from).

> Weird though that I can't reproduce this build failure, but I could the other. Anyways, hopefully that's also fixed now.

Sadly not :(
I am using zig 0.13 instead of master though, maybe that's a part of it? I 1) cloned the 0.13 branch 2) swapped `main.zig` in the `examples` directory with the websocket example at the bottom of the readme and 3) updated `build.zig.zon` with the new websocket dependency (since you only committed the update to master). Command is `zig build run -Dtarget=arm-linux-gnueabihf`.
Maybe I'll try a master build of zig and see if that changes anything. The application I plan on using this in uses 0.13 and not master which is why I've been using it so far.

```
http.zig [zig-0.13●] zig build run -Dtarget=arm-linux-gnueabihf
run
└─ run http.zig demo
   └─ zig build-exe http.zig demo Debug arm-linux-gnueabihf 1 errors
/home/myles/.cache/zig/p/1220db1a07a0ce926a10555705fb81dbe669477b4c3d9a6172bc885e5078eb2853e3/src/reader.zig:107:32: error: expected type 'usize', found 'u64'
    if ((try self.read(stream, data_needed)) == false) {
                               ^~~~~~~~~~~
/home/myles/.cache/zig/p/1220db1a07a0ce926a10555705fb81dbe669477b4c3d9a6172bc885e5078eb2853e3/src/reader.zig:107:32: note: unsigned 32-bit int cannot represent all possible unsigned 64-bit values
/home/myles/.cache/zig/p/1220db1a07a0ce926a10555705fb81dbe669477b4c3d9a6172bc885e5078eb2853e3/src/reader.zig:232:51: note: parameter type declared here
 fn read(self: *Reader, stream: anytype, to_read: usize) !bool {
                                                  ^~~~~
referenced by:
    readLoop__anon_11869: /home/myles/.cache/zig/p/1220db1a07a0ce926a10555705fb81dbe669477b4c3d9a6172bc885e5078eb2853e3/src/server.zig:298:38
    handle__anon_10934: /home/myles/.cache/zig/p/1220db1a07a0ce926a10555705fb81dbe669477b4c3d9a6172bc885e5078eb2853e3/src/server.zig:182:16
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
error: the following command failed with 1 compilation errors:
/usr/bin/zig build-exe -ODebug -target arm-linux-gnueabihf -mcpu baseline --dep httpz --dep metrics --dep websocket -Mroot=/home/myles/Documents/http.zig/example/main.zig --dep metrics --dep websocket --dep build -Mhttpz=/home/myles/Documents/http.zig/src/httpz.zig -Mmetrics=/home/myles/.cache/zig/p/1220232ab38d0c2cfb10680115c17ad2fa1a8531dbaf8bbfb359ec67e80c7d5f5758/src/metrics.zig -Mwebsocket=/home/myles/.cache/zig/p/1220db1a07a0ce926a10555705fb81dbe669477b4c3d9a6172bc885e5078eb2853e3/src/websocket.zig -Mbuild=/home/myles/Documents/http.zig/.zig-cache/c/0a991cbb44bf1c4520ed99cc54c5dead/options.zig --cache-dir /home/myles/Documents/http.zig/.zig-cache --global-cache-dir /home/myles/.cache/zig --name http.zig demo --listen=- 
Build Summary: 1/6 steps succeeded; 1 failed (disable with --summary none)
run transitive failure
└─ run http.zig demo transitive failure
   ├─ zig build-exe http.zig demo Debug arm-linux-gnueabihf 1 errors
   └─ install transitive failure
      └─ install http.zig demo transitive failure
         └─ zig build-exe http.zig demo Debug arm-linux-gnueabihf (+1 more reused dependencies)
error: the following build command failed with exit code 1:
/home/myles/Documents/http.zig/.zig-cache/o/7bb2a3422b4107402c0780ffbebb22b6/build /usr/bin/zig /home/myles/Documents/http.zig /home/myles/Documents/http.zig/.zig-cache /home/myles/.cache/zig --seed 0xf3ff08fd -Ze3ce3377deaf034e run -Dtarget=arm-linux-gnueabihf
```

**Comment by karlseguin** [comment]
Ok, I was able to get the build error. So I'm more confident that it's actually fixed now (it's in the 0.13 branch).

 I could change everything to u64, and that would work, because who's sending websocket messages > 4GB?  But I think it's more correct to keep it `data_needed: usize` and `payload_length: u64` AND to add a check that payload_length doesn't exceed what the architecture supports and turn it into a runtime error.

With a reasonable `max_size`, this never would have actually happened. The payload size check happens before any of these additions/allocations happen. But of course, the compiler doesn't know that and can't assume a "reasonble" `max_size`.

**Comment by MylesAndMore** [comment]
Great solution. Hopefully I won't need to send any 5GB messages over websockets anytime soon.

Everything compiles now and seems to work in my (limited) testing.
The nonblocking branch looks cool--although for now I will probably just deal with the memory leak as it doesn't really matter in my case and I don't want to migrate my project to zig master.

Thanks again for your help and sorry about the trouble! Who knew 32 less bits could cause so many issues...
